Submariner doc architecture overhaul stolostron/backlog#27003 (#4465)
open-cluster-management/rhacm-docs: add-ons/submariner/deploy_subm_manual.adoc

[#preparing-submariner-manual-deployment]

> = Deploying Submariner hosts manually

Reviewer: ```suggestion
= Deploying Submariner manually
```

> Before you deploy Submariner with {product-title}, you must prepare the clusters on the hosting environment for the connection. See xref:../submariner/submariner_prepare_hosts.adoc#preparing-selected-hosts-to-deploy-submariner[Deploying Submariner with the console] to learn how to automatically deploy Submariner on supported clusters by using the console.

Reviewer: ```suggestion
Before you deploy Submariner with {product-title}, you must prepare the clusters on the hosting environment for the connection. See xref:../submariner/submariner_prepare_hosts.adoc#preparing-selected-hosts-to-deploy-submariner[Deploying Submariner using? the console] to learn how to automatically deploy Submariner on supported clusters by using the console.
```

Reviewer 2: by using (otherwise it looks like submariner is using the console)

Author: Can we say "with the console" since we seem to use that in our docs? Or should we change to "by using"?

Reviewer 2: Well, I think it can be read as:

Deploying Submariner with the console vs Deploying Submariner without the console... 

But we are using the console to deploy submariner, which is prob the reason for the change request. 

Per IBM Style, we should be writing "by using" for that clarity, so you may not find that consistency in the doc...but it is the goal.

> If your cluster does not support automatic deployment, see the following topics to prepare your selected hosting environment manually for deployment on Submariner. Each host has unique steps for preparation, so make sure to select the correct host.

Reviewer: ```suggestion
If your cluster is hosted on a provider that does not support automatic Submariner deployment, see the following sections to prepare the infrastructure manually. Each provider has unique steps for preparation, so make sure to select the correct one.
```

[#preparing-ibm-submariner]
== Preparing IBM Cloud for Submariner

There are two kinds of Red Hat OpenShift Kubernetes Service (ROKS) on IBM Cloud: the classic cluster and the second generation of compute infrastructure in a virtual private cloud (VPC). Submariner cannot run on the classic ROKS cluster since it cannot configure the IPsec ports for the classic cluster.

> To configure the ROKS clusters on a VPC to use Submariner, complete the steps in the following links:

Reviewer: ```suggestion
To configure the ROKS clusters on a VPC to use Submariner, complete the following steps:
```

> . Before you create a cluster, specify subnets for pods and services, which avoids overlapping CIDRs with other clusters. Make sure there are no overlapping pods and services CIDRs between clusters if you are using an existing cluster. See https://cloud.ibm.com/docs/openshift?topic=openshift-vpc-subnets#vpc_basics[VPC Subnets] for the procedure.

Reviewer: ```suggestion
. If you plan to install Submariner without Globalnet, make sure that pod and service CIDRs do not overlap between clusters in the clusterSet. See https://cloud.ibm.com/docs/openshift?topic=openshift-vpc-subnets#vpc_basics[VPC Subnets] for the procedure. For clusters with overlapping CIDRs, Globalnet can be enabled while installing Submariner.
```

. Attach a public gateway to subnets used in the cluster. See https://cloud.ibm.com/docs/openshift?topic=openshift-vpc-subnets#vpc_basics_pgw[Public Gateway] for the procedure.

> . Create inbound rules for the default security group of the cluster by completing the steps in https://cloud.ibm.com/docs/openshift?topic=openshift-vpc-network-policy#security_groups_ui[Security Group]. Ensure that the firewall allows inbound and outbound traffic on 4500/UDP and 500/UDP ports for Gateway nodes, and allows inbound and outbound UDP/4800 for all the other nodes.

Reviewer: ```suggestion
. Create inbound rules for the default security group of the cluster by completing the steps in https://cloud.ibm.com/docs/openshift?topic=openshift-vpc-network-policy#security_groups_ui[Security Group]. Ensure that the firewall allows inbound and outbound traffic for external clients on 4500/UDP, 4490/UDP ports for the Gateway nodes, and allows inbound and outbound UDP/4800 traffic within the local cluster between the nodes.
```

> . Create inbound rules for the default security group of the cluster by completing the steps in https://cloud.ibm.com/docs/openshift?topic=openshift-vpc-network-policy#security_groups_ui[Security Group]. Ensure that the firewall allows inbound and outbound traffic for external clients on 4500/UDP, 4490/UDP ports for the Gateway nodes, and allows inbound and outbound UDP/4800 traffic within the local cluster between the nodes.

Reviewer: ```suggestion
. Create inbound rules for the default security group of the cluster by completing the steps in https://cloud.ibm.com/docs/openshift?topic=openshift-vpc-network-policy#security_groups_ui[Security Group]. Ensure that the firewall allows inbound/outbound traffic for external clients on 4500/UDP, 4490/UDP ports for the Gateway nodes, and allows inbound/outbound UDP/4800 traffic within the local cluster nodes.
```

> . Label a node that has the public gateway as `submariner.io/gateway=true` in the cluster.

Reviewer: ```suggestion
. Label the node selected as Gateway node in the previous step with the following label `submariner.io/gateway=true`.
```

. Refer to https://submariner.io/operations/deployment/calico/[Calico] to configure Calico CNI by creating IPPools in the cluster.

[#preparing-bare-metal-submariner]
== Preparing bare metal for Submariner

To prepare bare metal clusters for deploying Submariner, complete the following steps:

> . Ensure that the IPsec NATT and VXLAN ports are open.

Reviewer: ```suggestion
. Ensure that inbound and oubound VxLAN (UDP/4800) traffic between the cluster nodes is allowed in the firewall configuration.
```

Reviewer: At this step, we do not know the Gateway node, so lets add it as an extra step below.

> . Ensure that inbound and oubound VxLAN (UDP/4800) traffic between the cluster nodes is allowed in the firewall configuration.

Reviewer: ```suggestion
. Ensure that the firewall allows inbound/outbound traffic for external clients on 4500/UDP, 4490/UDP ports for the Gateway nodes, and allows inbound/outbound UDP/4800 traffic within the local cluster nodes.
```

. Customize and apply YAML content that is similar to the following example:

+
[source,yaml]
----
apiVersion: submarineraddon.open-cluster-management.io/v1alpha1
kind: SubmarinerConfig
metadata:
    name: submariner
    namespace: <managed-cluster-namespace>
spec:{}
----
+

> Replace `managed-cluster-namespace` with the namespace of your managed cluster. The name of the `SubmarinerConfig` must be `submariner`, as shown in the example.

Reviewer: ```suggestion
Replace `managed-cluster-namespace` with the name of your managed cluster. The name of the `SubmarinerConfig` must be `submariner`, as shown in the example.
```

+

> This configuration uses the default network address translation - traversal (NATT) port (4500/UDP) for your Submariner and one worker node is labeled as the Submariner gateway on your bare metal cluster.

Reviewer: ```suggestion
This configuration labels one of the worker node as the Submariner gateway on your bare metal cluster.
```

+

> Submariner uses IP security (IPsec) to establish the secure tunnels between the clusters on the gateway nodes. You can either use the default IPsec NATT port, or you can specify a different port that you configured. When you run this procedure without specifying an IPsec NATT port of 4500/UDP is automatically used for the communication.

Reviewer: ```suggestion
By default, Submariner uses IP security (IPsec) to establish the secure tunnels between the clusters on the gateway nodes. You can either use the default IPsec NATT port, or you can specify a different port that you configured. When you run this procedure without specifying an IPsec NATT port, 4500/UDP is used for the connections.
```

See xref:../submariner/submariner_customizations.adoc#submariner-customizations[Customizing Submariner deployments] for information about the customization options. 

> 

Reviewer: Please add this step 

. Identify the Gateway node configured by Submariner and enable firewall configuration to allow the IPsec NATT (UDP/4500), NatDiscovery (UDP/4490) ports for external traffic.

Author: Added

+
----
oc config view --flatten=true > rosa_kube/kubeconfig
----

. Import your OpenShift Service on AWS cluster to your cluster list by selecting *Infrastructure* > *Clusters* > *Import cluster* from the {product-title-short} console.

. Select the `Kubeconfig` *Import mode* and enter the content from your `kubeconfig` file in the *Kubeconfig* window. Follow the instructions in the console to complete the import.
+
You can verify that your OpenShift Service on AWS cluster was imported successfully by navigating to *Infrastructure* > *Clusters*.

. Navigate to *Infrastructure* > *Clusters* > *Cluster sets* and select the name of the cluster set that you want to add. Then, click the *Submariner add-ons* tab. 

. Click the *Install Submariner add-ons* button and set your OpenShift Service on AWS cluster as your *Target clusters*. Follow the instructions in the console to complete the installation.

. Navigate to *Infrastructure* > *Clusters* > *Cluster sets* > *Submariner add-ons* to verify that your OpenShift Service on AWS cluster *Connection status* is `Healthy`.

[#deploying-submariner-mcaddon-api]
== Deploy Submariner with the ManagedClusterAddOn API

> After manually preparing your selected hosting environment, you can deploy Submariner with the `ManagedClusterAddOn` API by complete the following steps:

Reviewer: ```suggestion
After manually preparing your selected hosting environment, you can deploy Submariner with the `ManagedClusterAddOn` API by completing the following steps:
```

. Create a `ManagedClusterSet` resource on the hub cluster by using the instructions provided in the _Creating and managing ManagedClusterSets_ topic of the link:../../multicluster_engine/cluster_lifecycle/managedclustersets_intro.adoc#managedclustersets_intro[Creating and managing ManagedClusterSets] documentation. Make sure your entry for the `ManagedClusterSet` resembles the following content:
+
[source,yaml]
----   
apiVersion: cluster.open-cluster-management.io/v1beta2
kind: ManagedClusterSet
metadata:
  name: <managed-cluster-set-name>
----
+
Replace `managed-cluster-set-name` with a name for the `ManagedClusterSet` that you are creating.
+

> *Note:* The maximum length of the name of the Kubernetes namespace is 63 characters, so the maximum length of the `<managed-cluster-set-name>` is 56 characters. If the length of `<managed-cluster-set-name>` exceeds 56, the `<managed-cluster-set-name>` is cut off from the head.

Reviewer: ```suggestion
*Note:* The maximum length of a Kubernetes namespace can be 63 characters. It is recommended that the maximum length of the `<managed-cluster-set-name>` is 56 characters or less. If the length of `<managed-cluster-set-name>` exceeds 56 chars, the `<managed-cluster-set-name>` is cut off from the head.
```

Author: As per IBM style, "can" indicates an ability. I'd consider the character limit more of a limit than an ability. I also wouldn't say that 56 characters or fewer is a recommendation if it's also a hard limitation. I'll make a couple changes to the wording but I think we need to keep most of it as is.

+
After the `ManagedClusterSet` is created, the `submariner-addon` creates a namespace called `<managed-cluster-set-name>-broker` and deploys the Submariner broker to it.

. Create the `Broker` configuration on the hub cluster in the `<managed-cluster-set-name>-broker` namespace by customizing and applying YAML content that is similar to the following example:
+
[source,yaml]
----
apiVersion: submariner.io/v1alpha1
kind: Broker
metadata:
     name: submariner-broker
     namespace: <managed-cluster-set-name>-broker
     labels:
         cluster.open-cluster-management.io/backup: submariner
spec:

>      globalnetEnabled: false

Reviewer: ```suggestion
     globalnetEnabled: <true-or-false>
```

----
+
Replace `managed-cluster-set-name` with the name of the managed cluster. 
+
Set the the value of `globalnetEnabled` to `true` if you want to enable Submariner Globalnet in the `ManagedClusterSet`.

. Add one managed cluster to the `ManagedClusterSet` by running the following command:
+
----
oc label managedclusters <managed-cluster-name> "cluster.open-cluster-management.io/clusterset=<managed-cluster-set-name>" --overwrite
----
+
Replace `<managed-cluster-name>` with the name of the managed cluster that you want to add to the `ManagedClusterSet`.
+
Replace `<managed-cluster-set-name>` with the name of the `ManagedClusterSet` to which you want to add the managed cluster. 

. Customize and apply YAML content that is similar to the following example:
+
[source,yaml]
----
apiVersion: submarineraddon.open-cluster-management.io/v1alpha1
kind: SubmarinerConfig
metadata:
    name: submariner
    namespace: <managed-cluster-namespace>
spec:{}
----
+

> Replace `managed-cluster-namespace` with the namespace of your managed cluster.

Reviewer: A User may not be aware that we create a namespace for every managed cluster. Since a namespace is created with the same name as the managed cluster, shall we rephrase the sentence as below?

CC: @nyechiel @aswinsuryan @vthapar 

```suggestion
Replace `managed-cluster-namespace` with the name of your managed cluster.
```

+
[source,yaml]
----
apiVersion: submarineraddon.open-cluster-management.io/v1alpha1
kind: SubmarinerConfig
metadata:
    name: submariner
    namespace: <managed-cluster-namespace>
spec:{}
----
+
Replace `managed-cluster-namespace` with the name of your managed cluster. The name of the `SubmarinerConfig` must be `submariner`, as shown in the example.
+
This configuration labels one of the worker nodes as the Submariner gateway on your bare metal cluster.
+
By default, Submariner uses IP security (IPsec) to establish the secure tunnels between the clusters on the gateway nodes. You can either use the default IPsec NATT port, or you can specify a different port that you configured. When you run this procedure without specifying an IPsec NATT port, 4500/UDP is used for the connections.

. Identify the Gateway node configured by Submariner and enable firewall configurations to allow the IPsec NATT (UDP/4500) and NatDiscovery (UDP/4490) ports for external traffic.

See xref:../submariner/submariner_customizations.adoc#customizing-submariner[Customizing Submariner deployments] for information about the customization options. 

[#preparing-aro-subariner]

> == Preparing Microsoft Azure Red Hat OpenShift for Submariner (Technology Preview)

Reviewer 3: As far as I know ARO is not a tech preview and going to be fully supported in the upcoming release.
So need to remove the tech preview note.

@nyechiel ^^

Author: If it is fully supported we will need to update the support matrix as well, and one or two sections in this doc.

Reviewer: I'll let @nyechiel comment about it, but AFAIR both ARO/ROSA (which use SM LB mode) are tech preview.
